### DISCUSSION TITLE
View XML table

### DISCUSSION THREAD

**Discussion by daviewales** [saulpw/visidata#1729] [Q&A]
I've got a big XML file from here:
http://www.mbsonline.gov.au/internet/mbsonline/publishing.nsf/Content/996BD3401470EC7ACA2588FD00152EC2/$File/MBS-XML-20230301-v2.XML

When I open it in Visidata, it looks like this:
![image](https://user-images.githubusercontent.com/2499066/217402728-13bd9d11-3f21-44c9-9be8-18dd0c491bc9.png)

However, the structure of the XML is like this:

``` xml
<MBS_XML>
  <Data>
    <ItemNum>3</ItemNum>
    <SubItemNum/>
    <ItemStartDate>01.12.1989</ItemStartDate>
    ...
  <Data>
    <ItemNum>4</ItemNum>
    <SubItemNum/>
    <ItemStartDate>01.12.1989</ItemStartDate>
    ...
  <Data>
    <ItemNum>23</ItemNum>
    <SubItemNum/>
    <ItemStartDate>01.12.1989</ItemStartDate>
    ...
</MBS_XML>
```

So ideally, I'd like to use each `<Data>` section as a row, and each item inside `<Data>` as a column.
Is there an easy way to do this?

**Comment by saulpw** [comment]
Hey @daviewales, I've been playing around with this a bit.  I think there are two parts:

1. selecting which nodes to include on the sheet
2. adding all relevant columns for those nodes

(1) is actually pretty straightforward; I would add an option `xml_xpath` and then change the `XmlSheet.iterload()` to iterate through `self.root.iterfind(self.options.xml_xpath)` and yield those.  This works for me with your example data (thanks for the use case; very helpful).

(2) is a bit trickier.  I made a simple XpathColumn:

```
class XpathColumn(Column):
    def calcValue(self, r):
        n = r.find(self.expr) if self.expr else r
        
        if n is not None:
            if self.attrib == 'text()':
                return ''.join(n.itertext()).strip()
            else:
                return r.attrib.get(self.attrib)
```

Then I added some code to the iterload loop:

```
        for elem in self.root.iterfind(self.options.xml_xpath):
            for subelem in elem.iter():
                k = unns(subelem.tag)
                if k not in self.xpathcols:
                    c = XpathColumn(k, expr=xpath_from(elem, subelem), attrib='text()')
                    self.addColumn(c)
                    self.xpathcols[k] = c

                for attrname in subelem.attrib:
                    k = unns(subelem.tag)
                    if k not in self.xpathcols:
                        c = XpathColumn(f'{k}_{attrname}', expr=k, attrib=attrname)
                        self.addColumn(c)
                        self.xpathcols[k] = c
```

Basically we want to add columns for each attribute of each node, and its text element, and iterate through all children.  The above code doesn't work yet (it at least needs `xpath_from` which should compute a viable xpath from the start node to the target node), but should be a start if you or someone wants to bang on it.

**Reply by daviewales** [reply]
Thanks! I'll try to remember to come back to this after I've finished the JSONL stuff.

I'm imagining a generic interface for tree-like data, similar to what happens when you open a JSON file with a tree structure.

1. When you first open the file, you see a single row with the root of the tree. In this example it would be: `<MBS_XML>`
2. When you press Enter, you dive down to the next level. I am imagining you would see lots of rows containing `<MBS_XML.Data>`
3. By default, when you dive into a single row, you see all the elements under that branch of the tree as rows.
4. However, perhaps there could be a command which says 'use the current level of the tree as the rows, and the next level as the columns'? So, when you get down to the 'Data' level, where you have rows of  `<MBS_XML.Data>`, you would execute the command, and it would push a new sheet where each column is `<MBS_XML.Data.Item>` and each value is the value of the item from that row and column.

I'm trying to think how to make this more generic, rather than needing to set an option to trigger it.
But maybe XML is too flexible for a generic approach.

**Reply by saulpw** [reply]
I like how you're thinking, and this could be a reasonable workflow for dealing with XML files.  In general I prefer 'nodes at all depths' to 'immediate children', because then you can use standard search, freq, whatever to find the kinds of nodes you're looking for.  But once you have those, we should have an easy way to push a new sheet with only those kinds of nodes (whether by xpath or tag), and then add the columns for all attributes and text content.

**Reply by daviewales** [reply]
Would it make sense to be able to switch between a flat all nodes view (current XML default) and a single level view (current JSON default), for _both_ XML and JSON? And would it make sense to consistently open into one of these for both?

**Reply by saulpw** [reply]
Hm, possibly.  Let's get it working for one format first, and then we can see how easily it translates.

**Comment by frosencrantz** [comment]
Occasionally I want to look at XML data.  I have some tools that generate simple XML data, not too deep.  I've found the XML sheet harder to look at data than JSON data, too.  One thing I've done is convert XML to JSON with https://github.com/martinblech/xmltodict, and I've found that data is easier to explore with Visidata and clean with jq.

I  wish there was an easier way to convert between equivalent formats, like this xmltodict or like json to gron and back within Visidata.

**Reply by aborruso** [reply]
@daviewales for XML files I do like @frosencrantz : I convert them to jsonlines using the great [yq/xq](https://github.com/kislyuk/yq) and I pass the output as jsonl to vd

```
curl 'http://www.mbsonline.gov.au/internet/mbsonline/publishing.nsf/Content/996BD3401470EC7ACA2588FD00152EC2/$File/MBS-XML-20230301-v2.XML' --compressed --insecure | \
xq -c '.MBS_XML.Data[]' | \
vd -f jsonl
```

![image](https://user-images.githubusercontent.com/30607/221401130-6cc7e24c-1fb3-4168-8ab1-21516f238186.png)
